test: Introduce ensure_for helper

fjahr opened the pull request:
A repeating pattern in the functional tests is that the test sleeps for a while to ensure that a certain condition is still true after some amount of time has elapsed. Most recently a new case of this was added in #30807. This PR here introduces an `ensure` helper to streamline this functionality.

Some approach considerations:
- It is possible to construct this by reusing `wait_until` and wrapping it in `try` internally. However, the logger output of the failing wait would still be printed which seems irritating. So I opted for simplified but similar internals to `wait_until`.
- This implementation starts for a failure in the condition right away which has the nice side-effect that it might give feedback on a failure earlier than is currently the case. However, in some cases, it may be expected that the condition may still be false at the beginning and then turns true until time has run out, something that would work when the test sleeps without checking in a loop. I decided against this design (and even against adding it as an option) because such a test design seems like it would be racy either way.
- I have also been going back and forth on naming. To me `ensure` works well but I am also not a native speaker, happy consider a different name if others don't think it's clear enough.

fjahr committed:
test: Introduce ensure_for helper

jonatack reviewed (commented):
Concept ACK

fjahr commented on test/functional/test_framework/util.py:
```diff
@@ -267,6 +267,14 @@ def satoshi_round(amount: Union[int, float, str], *, rounding: str) -> Decimal:
     return Decimal(amount).quantize(SATOSHI_PRECISION, rounding=rounding)
 
+def ensure_helper_internal(predicate, timeout, timeout_factor=1.0):
+    timeout = timeout * timeout_factor
```
My thought here was that we set a higher factor when things are generally expected to run slower, so in that case it seems natural to me that we will also want to wait longer. But yeah, I didn't take into account that people will set this factor unreasonably high to not see timeouts at all. I have removed it for now but I think an alternative could be to set a reasonable upper bound like 10x.

fjahr commented on test/functional/feature_assumeutxo.py:
```diff
@@ -305,8 +304,7 @@ def test_sync_from_assumeutxo_node(self, snapshot):
         # ibd_node's perceived unresponsiveness.
-        time.sleep(3)  # Sleep here because we can't detect when a node avoids requesting blocks from other peer.
-        assert_equal(len(ibd_node.getpeerinfo()[0]['inflight']), 0)
+        ibd_node.ensure(lambda: len(ibd_node.getpeerinfo()[0]['inflight']) == 0, 3)
```
Done and renamed to `duration`.

fjahr commented on test/functional/test_framework/p2p.py:
```diff
@@ -587,6 +588,9 @@ def test_function():
 
         wait_until_helper_internal(test_function, timeout=timeout, lock=p2p_lock, timeout_factor=self.timeout_factor)
 
+    def ensure(self, test_function, timeout=5):
```
sounds good, done

fjahr commented:
Addressed feedback from @maflcko and @jonatack and fixed the missing `lambda:` causing the CI failure.

fjahr commented on test/functional/test_framework/util.py:
```diff
@@ -267,6 +267,13 @@ def satoshi_round(amount: Union[int, float, str], *, rounding: str) -> Decimal:
     """Rounds a Decimal amount to the nearest satoshi using the specified rounding mode."""
     return Decimal(amount).quantize(SATOSHI_PRECISION, rounding=rounding)
 
+def ensure_helper_internal(predicate, timeout):
```
I didn't think it matters much because this is internal anyway, but I renamed it now as well.

fjahr commented on test/functional/test_framework/p2p.py:
```diff
@@ -587,6 +588,9 @@ def test_function():
 
         wait_until_helper_internal(test_function, timeout=timeout, lock=p2p_lock, timeout_factor=self.timeout_factor)
 
+    def ensure(self, test_function, *, duration=5):
```
Ok, that might be more of a naming issue with `ensure` then which I am happy to reconsider if someone has a better idea but I removed the default value for now.

fjahr replied:
Maybe rather `ensure_for(duration=X...`? "after" implies that the condition will only be checked once after time is up IMO...

BrandonOdiwuor reviewed (commented):
Concept ACK

pablomartin4btc reviewed (commented):
Concept ACK

nit:
> - I have also been going back and forth on naming.

What about `ensure_during`? (another alternatives: `ensure_within_duration`/ `ensure_over_duration`) It adds a notion of time, making it more specific to the context of waiting or checking a condition over a period and it's less generic and more intuitive.

fjahr commented:
Since a few reviewers seem to prefer a clearer naming, I have now changed it to `ensure_for(duration=..., f=...)`.

furszy commented on test/functional/test_framework/util.py:
```diff
@@ -267,6 +267,13 @@ def satoshi_round(amount: Union[int, float, str], *, rounding: str) -> Decimal:
     """Rounds a Decimal amount to the nearest satoshi using the specified rounding mode."""
     return Decimal(amount).quantize(SATOSHI_PRECISION, rounding=rounding)
 
+def ensure_for_helper_internal(*, duration, predicate):
+    time_end = time.time() + duration
+    predicate_source = "''''\n" + inspect.getsource(predicate) + "'''"
+    while time.time() < time_end:
+        if not predicate():
+            raise AssertionError("Predicate {} became false within {} seconds".format(predicate_source, duration))
+        time.sleep(0.05)
```
Isn't it a bit aggressive to execute the provided predicate every 0.05 seconds for the entire duration?
Beyond flooding the logs, the predicate likely calls an RPC command that locks `cs_main` or some other important mutex, which slows down the node.

For instance, the `feature_assumeutxo.py` one, will call `getpeerinfo()` 60 times.

fjahr replied:
I simply took the same interval as `wait_until` uses. If the slowdown was an issue it should be much more apparent from the use there, so I did a simple benchmark and increasing the sleep from `0.05` to `0.2` lead to a considerable slowdown of the test suite overall (8min to 10:45min), so I don't think this would be an issue here either.

We could still lower the interval. Since we generally don't expect the predicate to become false it shouldn't hurt performance as much as changing it in `wait_until`. It means we would have fewer logs and if a failure occurs it would take a bit longer to detect it.

fjahr replied:
> And, in fact, we could implement it using the wait_until function too

Yes, I tried this and decided against it, see the description of the PR.

> I’d suggest simply waiting for the entire duration without intermediate checks.

I have also considered this and decided against it because I don't think it's better. I have also described this in the PR description already. Since the polling doesn't seem to lead to a real performance issue I think it's better to fail faster if we fail. This also uncovers potential races, as mentioned in the description.

furszy replied:
> > And, in fact, we could implement it using the wait_until function too
> 
> Yes, I tried this and decided against it, see the description of the PR.

Ah, missed it. Sorry.

> > I’d suggest simply waiting for the entire duration without intermediate checks.
> 
> I have also considered this and decided against it because I don't think it's better. I have also described this in the PR description already. Since the polling doesn't seem to lead to a real performance issue I think it's better to fail faster if we fail. This also uncovers potential races, as mentioned in the description.

hmm, so you are thinking about a condition that passes at the beginning, does not passes somewhere during the test sleep time and ends up passing at the end. Maybe the test condition should be changed if such scenario exist?

I'm not sure we can really ensure that this won't lead to performance issues. It's abnormal to fail here, as it's something we don't expect during the test. E.g. we would be making 59 additional RPC calls in `feature_assumeutxo.py`, which would flood the active logs (logs we may need if the test fails later for some other reason after wise), all while looking for a failure that will rarely occur.
I mean, for current use cases, it probably doesn't matter. But could this be an issue for future uses? It seems to be something it could be misused very easily.

fjahr replied:
> hmm, so you are thinking about a condition that passes at the beginning, does not passes somewhere during the test sleep time and ends up passing at the end. Maybe the test condition should be changed if such scenario exist?

Yes, exactly, and in order to know that it should be changed the person writing the test will see that `ensure_for` fails. This is the protection against introducing flakiness that I mentioned.

> which would flood the active logs

I don't know if that's a problem because I always only grep the logs anyway.

> I'm not sure we can really ensure that this won't lead to performance issues.

The benchmarks say no but if you have another way to show that there is a performance impact. Maybe that's not a guarantee but if there is any it appears to be very small.

However, I have made the delay configurable now so this can be adjusted and I have set the default to 0.2 instead of 0.05. That should keep impact on the logs (and performance) small. I hope that works for everyone.

tdb3 reviewed (approved):
ACK dcd66472600f1676e2e450245924ab6e1a837d39

Nice addition.  Compliments `wait_until` nicely.

(As expected) saw little difference in both individual and total runtimes of tests on master vs pr branch (at least within typical run-to-run difference).

### Master (e46bebb444dfff89813333dc1926a57c9a3e8aab):
| test | status | duration(seconds) |
|---|---|---|
| feature_assumeutxo.py | Passed | 32 |
| feature_minchainwork.py | Passed | 10 |
| mempool_unbroadcast.py | Passed | 8 |
| p2p_segwit.py | Passed | 67 |

Total runtime: 141s

### PR branch:
| test | status | duration(seconds) |
|---|---|---|
| feature_assumeutxo.py   | Passed  | 32 |
| feature_minchainwork.py | Passed  | 10 |
| mempool_unbroadcast.py  | Passed  | 7 |
| p2p_segwit.py | Passed  | 67 |

Total runtime: 141s

fjahr committed:
test: Add check_interval parameter to wait_until

This also replaces two sleep calls in functional tests with wait_until

fjahr commented:
I have introduced a delay parameter that allows to adjust the delay between checking the predicate. The default is set at 0.2 instead of previously 0.05.

I have also added a second commit adding the delay parameter to `wait_until` and using it in places where a `time.sleep` is currently used and a high frequency of checking does not seem appropriate.

There are still a few sleeps spread across the tests, I think some of them require some further thought on how to remove them and some of them could probably be left alone. See for example #30942 that I found while working on this here but I am keeping it separate since we can get rid of the sleep there for different reasons without a replacement.

furszy commented on test/functional/test_framework/util.py:
```diff
@@ -267,6 +267,13 @@ def satoshi_round(amount: Union[int, float, str], *, rounding: str) -> Decimal:
     """Rounds a Decimal amount to the nearest satoshi using the specified rounding mode."""
     return Decimal(amount).quantize(SATOSHI_PRECISION, rounding=rounding)
 
+def ensure_for_helper_internal(*, duration, predicate, delay=0.2):
+    time_end = time.time() + duration
+    predicate_source = "''''\n" + inspect.getsource(predicate) + "'''"
+    while time.time() < time_end:
+        if not predicate():
+            raise AssertionError("Predicate {} became false within {} seconds".format(predicate_source, duration))
+        time.sleep(delay)
```
tiny nit: if `duration < delay`, then `delay` should be set to duration?

fjahr replied:
Hm, just doing that would have an unexpected effect, I think. With the code unchanged if the user would today pass the parameters with delay=duration, then I think the predicate would only be checked once in the beginning and not at the end because at the second conditional check the time passed should be the runtime of predicate + the sleep of duration, so the loop wouldn't run a second time. That's not what a user expects when they pass a high delay probably. So I have done the following: If delay is 0, negative or higher than duration, we set it equal to duration. But also the predicate is now check once in the beginning explicitly and in the loop it now runs after the sleep instead of before. This means that if the user sets delay to 0 explicitly (or set a value that is weird, negative or above duration) they will have the behavior or check once in the beginning and once at the end. This should be the minimum of checks, as discussed previously.

furszy commented on test/functional/test_framework/p2p.py:
```diff
@@ -580,13 +580,13 @@ def on_version(self, message):
     # Connection helper methods
 
-    def wait_until(self, test_function_in, *, timeout=60, check_connected=True):
+    def wait_until(self, test_function_in, *, timeout=60, check_connected=True, delay=0.05):
```
maybe rename "delay" to "check_interval" or another similar term?

fjahr replied:
renamed to `check_interval`

furszy reviewed (commented):
looking good, left few nits.

fjahr commented:
Addressed feedback from @furszy and added some better documentation.

furszy reviewed (commented):
Code review ACK 

Commit 19bc71caf6a43702946f1bc15696fe14afa24fda description says "delay" when it should say "check_interval".

fjahr commented:
> Commit https://github.com/bitcoin/bitcoin/commit/19bc71caf6a43702946f1bc15696fe14afa24fda description says "delay" when it should say "check_interval".

Also fixed.

furszy reviewed (commented):
ACK 5bcc578b7f5b34bd4c3a6ee1e54681c56e3db450

fjahr committed:
test: Remove unused attempts parameter from wait_until

fjahr commented on test/functional/wallet_keypool.py:
```diff
@@ -137,8 +136,7 @@ def run_test(self):
         # test walletpassphrase timeout
-        time.sleep(1.1)
-        assert_equal(nodes[0].getwalletinfo()["unlocked_until"], 0)
+        self.nodes[0].wait_until(lambda: nodes[0].getwalletinfo()["unlocked_until"] == 0, timeout=1.1, check_interval=0.5)
```
Yeah, looking at it again, this doesn't make sense because the test may fail if someone actually scales the wait time down. I changed it back and will leave the mocktime approach for a follow-up.

fjahr commented on test/functional/test_framework/test_node.py:
```diff
@@ -838,9 +839,11 @@ def bumpmocktime(self, seconds):
+        return wait_until_helper_internal(test_function, timeout=timeout, timeout_factor=self.timeout_factor, check_interval=check_interval)
 
+    def ensure_for(self, *, duration, f, check_interval=0.2):
+        return ensure_for_helper_internal(duration=duration, predicate=f, check_interval=check_interval)
```
Changed to use `**kwargs` in both `ensure_for` impls

fjahr commented on test/functional/test_framework/util.py:
```diff
@@ -268,7 +268,28 @@ def satoshi_round(amount: Union[int, float, str], *, rounding: str) -> Decimal:
+        time.sleep(check_interval)
+
+
+def wait_until_helper_internal(predicate, *, attempts=float('inf'), timeout=float('inf'), lock=None, timeout_factor=1.0, check_interval=0.05):
```
I agree and couldn't find any good use cases for it. It can be re-added when there is a use case for it so I have removed it for now and simplified the code.

fjahr commented:
Addressed feedback from @maflcko , thanks!

tdb3 reviewed (approved):
re ACK 352b8209aa5327f7d369e2acc4d87f9767389a6b

Also sanity checked testing timing (https://github.com/bitcoin/bitcoin/pull/30893?notification_referrer_id=NT_kwDOBljilbUxMjM4MjY5Nzc4OToxMDY0ODg0Njk#pullrequestreview-2316167311)

furszy reviewed (commented):
ACK 352b8209aa5

tdb3 reviewed (approved):
Code Review re ACK f04529fc40f2f7c8e4e63d892cdcbe6c91dfbf96

Would also review again if the suggestion in https://github.com/bitcoin/bitcoin/pull/30893#discussion_r1834183665 was implemented.

fjahr commented on test/functional/test_framework/test_node.py:
```diff
@@ -841,6 +842,8 @@ def bumpmocktime(self, seconds):
     def wait_until(self, test_function, timeout=60):
         return wait_until_helper_internal(test_function, timeout=timeout, timeout_factor=self.timeout_factor)
 
+    def ensure_for(self, *, duration, f, **kwargs):
```
Initially it had the `timeout_factor` and also it seemed nice to have it next to `wait_until` but sure, I have made it a util function now.

fjahr commented:
Addressed feedback from @maflcko and made `ensure_for` a util function.

tdb3 reviewed (approved):
code review re ACK dde84e2b2b6a64f5f67d4dff6e3fb2fc7d2c5a77

Changes are focused on the movement of `ensure_for()` to util (with appropriate adjustments to imports and calls to it).

maflcko reviewed (commented):
almost ack dde84e2b2b6a64f5f67d4dff6e3fb2fc7d2c5a77 🚿

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: almost ack dde84e2b2b6a64f5f67d4dff6e3fb2fc7d2c5a77 🚿
nzWmeb5moWyuyuQLS3fgHHElryIVOHlOhCoY+A+Ny0e4Eo+J09cUfv98NBfJtN3Huhk9sqVRYf7s8+PEHS/zCw==
```

</details>

fjahr commented on test/functional/wallet_encryption.py:
```diff
@@ -53,8 +54,7 @@ def run_test(self):
         # Check that the timeout is right
-        time.sleep(3)
-        assert_raises_rpc_error(-13, "Please enter the wallet passphrase with walletpassphrase first", self.nodes[0].signmessage, address, msg)
+        self.nodes[0].wait_until(lambda: try_rpc(-13, "Please enter the wallet passphrase with walletpassphrase first", self.nodes[0].signmessage, address, msg), timeout=3, check_interval=0.5)
```
What do you mean by "accurate"? Is the concern that someone scales this by <0.5 and then the test might fail?

fjahr replied:
reverted it

fjahr commented on test/functional/test_framework/util.py:
```diff
@@ -268,6 +268,27 @@ def satoshi_round(amount: Union[int, float, str], *, rounding: str) -> Decimal:
+    predicate_source = "''''\n" + inspect.getsource(f) + "'''"
+    while True:
+        if not f():
+            raise AssertionError("Predicate {} became false within {} seconds".format(predicate_source, duration))
```
done

fjahr commented on test/functional/test_framework/util.py:
```diff
@@ -268,6 +268,27 @@ def satoshi_round(amount: Union[int, float, str], *, rounding: str) -> Decimal:
     return Decimal(amount).quantize(SATOSHI_PRECISION, rounding=rounding)
 
 
+def ensure_for(*, duration, f, check_interval=0.2, **kwargs):
```
done

fjahr commented:
Addressed latest feedback from @maflcko

tdb3 reviewed (approved):
code review re ACK 111465d72dd35e42361fc2a089036f652417ed37

furszy reviewed (commented):
utACK 111465d72dd35e42361fc2a089036f652417ed37

achow101 commented:
ACK 111465d72dd35e42361fc2a089036f652417ed37
